[regression] java.lang.VerifyError: Operand stack overflow

Related issues
- https://github.com/redhat-developer/vscode-java/issues/3232
- https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1139

Steps to reproduce
- create the following class
```
public class Test {
	public static void main(String[] args) {
		String value = "abc";
		value.equalsIgnoreCase("" + null);
		System.out.println(value.substring(1));
	}
}
```
- run it
It throws 
```
Error: Unable to initialize main class org.sample.Test
Caused by: java.lang.VerifyError: Operand stack overflow
Exception Details:
  Location:
    org/sample/Test.main([Ljava/lang/String;)V @17: iconst_1
  Reason:
    Exceeded max stack size.
  Current Frame:
    bci: @17
    flags: { }
    locals: { '[Ljava/lang/String;', 'java/lang/String' }
    stack: { 'java/io/PrintStream', 'java/lang/String' }
  Bytecode:
    0000000: 1210 4c2b ba00 1200 00b6 0016 57b2 001c
    0000010: 2b04 b600 22b6 0026 b1                 

 ```
A workaorund
set `org.eclipse.jdt.core.compiler.codegen.useStringConcatFactory = disabled`

The issue has been caused by https://github.com/eclipse-jdt/eclipse.jdt.core/commit/be45d8f4905bb48bdce51f8adf2ed85e5089e22c

It can't be reproduced in Eclipse <= I20230628-1800

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Expression.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Expression.java
@@ -967,13 +967,12 @@ public void buildStringForConcatation(BlockScope blockScope, CodeStream codeStre
 				break;
 			default :
 				if (this.resolvedType.id == TypeIds.T_null) {
-					// Optimize it, avoid aconst_null, simply append the String Literal
-					recipe.append((String) null);
+					codeStream.aconst_null();
 				} else {
 					generateCode(blockScope, codeStream, true);
 					codeStream.invokeStringValueOf(typeID);
-					addArgumentToRecipe(blockScope, codeStream, recipe, blockScope.getJavaLangString(), argTypes);
 				}
+				addArgumentToRecipe(blockScope, codeStream, recipe, blockScope.getJavaLangString(), argTypes);
 				break;
 		}
 	} else {

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/ConstantTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/ConstantTest.java
@@ -378,7 +378,8 @@ public void test009() throws Exception {
 								+ "    36  invokevirtual java.lang." + substring1 + ".append(java.lang.String) : java.lang." + substring1 + " [44]\n"
 								+ "    39  invokevirtual java.lang." + substring1 + ".toString() : java.lang.String [47]\n"
 									:
-									"    21  invokedynamic 0 makeConcatWithConstants() : java.lang.String [32]\n";
+									"    21  aconst_null\n" +
+									"    22  invokedynamic 0 makeConcatWithConstants(java.lang.String) : java.lang.String [32]\n";
 	String expectedOutput =
 			"  // Method descriptor #15 ([Ljava/lang/String;)V\n" +
 			"  // Stack: 3, Locals: 4\n" +
@@ -446,44 +447,44 @@ public void test009() throws Exception {
 			"    16  ldc <String \"2\"> [30]\n" +
 			"    18  invokevirtual java.io.PrintStream.print(java.lang.String) : void [24]\n" +
 			substring2 +
-			  "    26  ldc <String \"[null]\"> [36]\n"
-			+ "    28  if_acmpne 35\n"
-			+ "    31  iconst_1\n"
-			+ "    32  goto 36\n"
-			+ "    35  iconst_0\n"
-			+ "    36  istore_1 [b]\n"
-			+ "    37  getstatic java.lang.System.out : java.io.PrintStream [16]\n"
-			+ "    40  ldc <String \"3\"> [38]\n"
-			+ "    42  invokevirtual java.io.PrintStream.print(java.lang.String) : void [24]\n"
-			+ "    45  aconst_null\n"
-			+ "    46  astore_2 [s]\n"
-			+ "    47  aload_2 [s]\n"
-			+ "    48  ifnonnull 59\n"
-			+ "    51  getstatic java.lang.System.out : java.io.PrintStream [16]\n"
-			+ "    54  ldc <String \"4\"> [40]\n"
-			+ "    56  invokevirtual java.io.PrintStream.print(java.lang.String) : void [24]\n"
-			+ "    59  ldc <String \"aaa\"> [42]\n"
-			+ "    61  astore_3 [s2]\n"
-			+ "    62  getstatic java.lang.System.out : java.io.PrintStream [16]\n"
-			+ "    65  ldc <String \"5\"> [44]\n"
-			+ "    67  invokevirtual java.io.PrintStream.println(java.lang.String) : void [46]\n"
-			+ "    70  return\n"
-			+ "      Line numbers:\n"
-			+ "        [pc: 0, line: 3]\n"
-			+ "        [pc: 8, line: 4]\n"
-			+ "        [pc: 13, line: 5]\n"
-			+ "        [pc: 21, line: 6]\n"
-			+ "        [pc: 37, line: 7]\n"
-			+ "        [pc: 45, line: 8]\n"
-			+ "        [pc: 47, line: 9]\n"
-			+ "        [pc: 59, line: 10]\n"
-			+ "        [pc: 62, line: 11]\n"
-			+ "        [pc: 70, line: 12]\n"
-			+ "      Local variable table:\n"
-			+ "        [pc: 0, pc: 71] local: args index: 0 type: java.lang.String[]\n"
-			+ "        [pc: 37, pc: 71] local: b index: 1 type: boolean\n"
-			+ "        [pc: 47, pc: 71] local: s index: 2 type: java.lang.String\n"
-			+ "        [pc: 62, pc: 71] local: s2 index: 3 type: java.lang.String\n";
+			"    27  ldc <String \"[null]\"> [36]\n" +
+			"    29  if_acmpne 36\n" +
+			"    32  iconst_1\n" +
+			"    33  goto 37\n" +
+			"    36  iconst_0\n" +
+			"    37  istore_1 [b]\n" +
+			"    38  getstatic java.lang.System.out : java.io.PrintStream [16]\n" +
+			"    41  ldc <String \"3\"> [38]\n" +
+			"    43  invokevirtual java.io.PrintStream.print(java.lang.String) : void [24]\n" +
+			"    46  aconst_null\n" +
+			"    47  astore_2 [s]\n" +
+			"    48  aload_2 [s]\n" +
+			"    49  ifnonnull 60\n" +
+			"    52  getstatic java.lang.System.out : java.io.PrintStream [16]\n" +
+			"    55  ldc <String \"4\"> [40]\n" +
+			"    57  invokevirtual java.io.PrintStream.print(java.lang.String) : void [24]\n" +
+			"    60  ldc <String \"aaa\"> [42]\n" +
+			"    62  astore_3 [s2]\n" +
+			"    63  getstatic java.lang.System.out : java.io.PrintStream [16]\n" +
+			"    66  ldc <String \"5\"> [44]\n" +
+			"    68  invokevirtual java.io.PrintStream.println(java.lang.String) : void [46]\n" +
+			"    71  return\n" +
+			"      Line numbers:\n" +
+			"        [pc: 0, line: 3]\n" +
+			"        [pc: 8, line: 4]\n" +
+			"        [pc: 13, line: 5]\n" +
+			"        [pc: 21, line: 6]\n" +
+			"        [pc: 38, line: 7]\n" +
+			"        [pc: 46, line: 8]\n" +
+			"        [pc: 48, line: 9]\n" +
+			"        [pc: 60, line: 10]\n" +
+			"        [pc: 63, line: 11]\n" +
+			"        [pc: 71, line: 12]\n" +
+			"      Local variable table:\n" +
+			"        [pc: 0, pc: 72] local: args index: 0 type: java.lang.String[]\n" +
+			"        [pc: 38, pc: 72] local: b index: 1 type: boolean\n" +
+			"        [pc: 48, pc: 72] local: s index: 2 type: java.lang.String\n" +
+			"        [pc: 63, pc: 72] local: s2 index: 3 type: java.lang.String\n";
 	if (this.complianceLevel >= ClassFileConstants.JDK9) {
 		int index = actualOutput.indexOf(expectedOutput9OrLater);
 		if (index == -1 || expectedOutput9OrLater.length() == 0) {
@@ -1621,6 +1622,21 @@ public void testBug569498() {
 				+ "}",
 			},
 			"");
+}
+//https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1256
+public void testGH1256() throws Exception {
+	this.runConformTest(
+		new String[] {
+			"X.java",
+			"public class X {\n"
+			+ "	public static void main(String[] args) {\n"
+			+ "		String value = \"12345\";\n"
+			+ "		value.equalsIgnoreCase(\"\" + null);\n"
+			+ "		System.out.println(value.substring(1));\n"
+			+ "	}\n"
+			+ "}",
+		 },
+	"2345");
 }
 public static Class testClass() {
 	return ConstantTest.class;

